File: lib/repository/comment/my_sql.py
from datetime import datetime;
from interfaces.comment_repository import ICommentRepository;

import os;
import uuid;
import mysql.connector;

# Implements ICommentRepository interface for connecting to a MySQL database.
# See interfaces/comment_repository for method documentation


class CommentMySQLRepository(ICommentRepository):
    _table_name = "comments";


    def __init__(self, field_map):
        """
        @param (object) self
        @param (dict) field_map - Map of fields returned for database queries returned from MySQL connector; used to create an instance of a specific class after the raw data is fetched from the database.
        @returns (None)
        """
        self._db_connection = mysql.connector.connect(
          host = os.getenv("DATABASE_HOSTNAME"),
          user = os.getenv("DATABASE_USERNAME"),
          password = os.getenv("DATABASE_PASSWORD"),
          database = os.getenv("DATABASE_NAME")
        );
        self._field_map = field_map;


    def create(self, doc):
        db_cursor = self._db_connection.cursor();
        my_uuid = str(uuid.uuid4());
        created_date = datetime.now();
        query = ("INSERT INTO comments (id, post_id, user_id, body, created_date) VALUES (%s, %s, %s, %s, %s)");

        doc.update({"id": my_uuid, "created_date": created_date});
        db_cursor.execute(query, (my_uuid, doc["post_id"], doc["user_id"], doc["body"], doc["created_date"]));
        self._db_connection.commit();
        db_cursor.close();

        return { "id": my_uuid, "created_date": str(created_date) };


    def find_one(self, id):
        db_cursor = self._db_connection.cursor();
        query = ("SELECT * FROM posts WHERE id = '{}'".format(id));

        db_cursor.execute(query);
        result = db_cursor.fetchall();

        return list(map(lambda c: self.on_read_comment(c), result));


    def find_all_comments(self):
        comment_list = []
        db_cursor = self._db_connection.cursor();
        db_cursor.execute("SELECT * FROM comments");
        result = db_cursor.fetchall();

        for comment in result:
          comment_list.append(self.on_read_comment(comment));

        return comment_list;


    def edit_comment(self, id, text):
        db_cursor = self._db_connection.cursor();
        last_modified = datetime.now();
        query = ("UPDATE comments SET body = '{}', last_modified = '{}' WHERE id = '{}'".format(text, id, last_modified));

        db_cursor.execute(query);
        self._db_connection.commit();
        db_cursor.close();

        return { "id": id, "last_modified": str(last_modified) };


    def incr_like_count(self, id):
        db_cursor = self._db_connection.cursor();
        query = ("UPDATE comments SET like_count = like_count + 1 WHERE id = '{}'".format(id));

        db_cursor.execute(query);
        self._db_connection.commit();
        db_cursor.close();


    def on_read_comment(self, record):
        return {
          "id": record[self._field_map["id"]],
          "post_id": record[self._field_map["post_id"]],
          "user_id": record[self._field_map["user_id"]],
          "body": record[self._field_map["body"]],
          "like_count": record[self._field_map["like_count"]],
          "created_date": record[self._field_map["created_date"]]
        }

####CommentMySQLRepository###

File: src/lib/repository/comment/mysql.js
/* istanbul ignore file */

/*Implements ICommentRepository interface for connecting to a MySQL database.
*See interfaces/comment-repository for method documentation
*/

const uuid = require("uuid");
const { promisify } = require("util");

/**
 * @implements {ICommentRepository}
 * @param {Object} databaseConnector - object with methods for connecting to a database 
 */

function CommentMySQLRepository(databaseConnector) {

    this.create = async function(doc) {
        const createdDate = new Date().toISOString();
        const connection = await databaseConnector.getConnection();
        const runQueryWith = promisify(connection.query.bind(connection));
        const id = uuid.v4();
        const sql = `INSERT INTO comments (id, post_id, user_id, body, created_date) VALUES ('${id}', '${doc.postId}', '${doc.userId}', '${doc.body}', '${createdDate}')`;
        const result = await runQueryWith(sql);
        connection.release();

        return { id, createdDate };
    }

    this.findOneById = async function(id) {
        const createdDate = new Date().toISOString();
        const connection = await databaseConnector.getConnection();
        const runQueryWith = promisify(connection.query.bind(connection));
        const sql = `SELECT * FROM comments WHERE id = '${id}'`;
        const result = await runQueryWith(sql);
        connection.release();

        return result.map((c) => onReadComment(c));
    }

    this.findAllComments = async function() {
        const createdDate = new Date().toISOString();
        const connection = await databaseConnector.getConnection();
        const runQueryWith = promisify(connection.query.bind(connection));
        const sql = `SELECT id, user_id, post_id, body, like_count, created_date FROM comments`
        const result = await runQueryWith(sql);
        connection.release();

        return result.map((c) => onReadComment(c));
    }


    this.editComment = async function(id, text) {
        const lastModified = new Date().toISOString();
        const connection = await databaseConnector.getConnection();
        const runQueryWith = promisify(connection.query.bind(connection));
        const sql = `UPDATE comments SET body = '${text}', last_modified = '${lastModified}' WHERE id = '${id}'`;

        const result = await runQueryWith(sql);
        connection.release();

        return { id, lastModified };
    }

    this.incrementLikeCount = async function({commentId, userId}) {
        const connection = await databaseConnector.getConnection();
        const runQueryWith = promisify(connection.query.bind(connection));
        const incrementLikeCountSql = `INSERT INTO comment_likes (comment_id, user_id) VALUES ("${commentId}", "${userId}")`;
        const getLikeCountSql = `SELECT COUNT(*) AS like_count FROM comment_likes WHERE comment_id = "${commentId}"`;
        const updateCommentLikeCountSql = `UPDATE posts SET like_count = {like_count} WHERE id = "${commentId}"`;
        const checkUserHasAlreadyLikedCommentSql = `SELECT COUNT(*) AS likes FROM comment_likes WHERE EXISTS
        (SELECT user_id FROM post_likes WHERE user_id = "${userId}" AND comment_id = "${commentId}")`;
        
        const [fromCurrentUser] = await runQueryWith(checkUserHasAlreadyLikedCommentSql);
        const userHasAlreadyLikedComment = fromCurrentUser.likes !== 0;

        if (userHasAlreadyLikedComment) {
            return;
        }

        connection.beginTransaction(async(err) => {
            try {
                await runQueryWith(incrementLikeCountSql);
                const [result] = await runQueryWith(getLikeCountSql);
                const {like_count} = result; 
                await runQueryWith(updateCommentLikeCountSql.replace("{like_count}", like_count));
                connection.commit();
                connection.release();

            }
            catch (e) {
                console.error(e);
                connection.rollback();
                connection.release();

            }
        });

    }

    this.decrementLikeCount = async function({commentId, userId}) {
        const connection = await databaseConnector.getConnection();
        const runQueryWith = promisify(connection.query.bind(connection));
        const decrementLikeCountSql = `DELETE FROM comment_likes WHERE comment_id = "${commentId}" AND user_id = "${userId}"`;
        const getLikeCountSql = `SELECT COUNT(*) AS like_count FROM comment_likes WHERE comment_id = "${commentId}"`;
        const updateCommentLikeCountSql = `UPDATE posts SET like_count = {like_count} WHERE id = "${commentId}"`;
        const checkUserHasAlreadyUnLikedCommentSql = `SELECT COUNT(*) AS likes FROM comment_likes WHERE EXISTS
        (SELECT user_id FROM post_likes WHERE user_id = "${userId}" AND comment_id = "${commentId}")`;
        
        const [fromCurrentUser] = await runQueryWith(checkUserHasAlreadyUnLikedCommentSql);
        const userHasAlreadyUnLikedComment = fromCurrentUser.likes === 0;

        if (userHasAlreadyUnLikedComment) {
            return;
        }

        connection.beginTransaction(async(err) => {
            try {
                await runQueryWith(decrementLikeCountSql);
                const [result] = await runQueryWith(getLikeCountSql);
                const {like_count} = result; 
                await runQueryWith(updateCommentLikeCountSql.replace("{like_count}", like_count));
                connection.commit();
                connection.release();
            }
            catch (e) {
                console.error(e);
                connection.rollback();
                connection.release();
            }
        });

    }


    function onReadComment(record) {
        return {
            id: record.id,
            postId: record.post_id,
            userId: record.user_id,
            body: record.body,
            likeCount: record.like_count,
            createdDate: record.created_date
        }
    }

}


/*CommentMySQLRepository*/

module.exports = CommentMySQLRepository;
